src/vala-previewer/previewer.vala

```diff
@@ -0,0 +1,111 @@
+            if (target == null) {
+                stderr.printf (@"Widget with target_id='$target_id' could not be found.\n");
+                return;
+            }
+            this.window.child = target;
+        }
+
+        public void update_css (string content) {
+            if (this.css != null)
+                Gtk.StyleContext.remove_provider_for_display (Gdk.Display.get_default (), this.css);
+            this.css = new Gtk.CssProvider ();
+            this.css.load_from_data (content.data);
+            Gtk.StyleContext.add_provider_for_display (Gdk.Display.get_default (), this.css , Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
+        }
+
+        // filename to loadable module or empty string ("") to just run it again
+        // also builder_symbol can be empty. Then the builder object is not handed over to the module
+        public void run (string filename, string run_symbol, string builder_symbol) {
+            if (filename == "") {
+                if (this.module == null) {
+                    stderr.printf ("No Module specified yet.\n");
+                    return;
+                }
+            } else {
+                if (!Module.supported ()) {
+                    stderr.printf ("This system does not support loadable modules.\n");
+                    return;
+                }
+
+                this.module.close ();
+                this.module = Module.open (filename, ModuleFlags.LAZY);
+                if (this.module == null) {
+                    stderr.printf ("Module loading failed.\n");
+                    return;
+                }
+            }
+
+            if (builder_symbol != "") {
+                if (this.builder == null) {
+                    stderr.printf ("No UI definition loaded yet.\n");
+                    return;
+                }
+
+                this.window.present ();
+                this.presented = true;
+
+                void* function;
+                this.module.symbol (builder_symbol, out function);
+                if (function == null) {
+                    stderr.printf (@"Module does not contain symbol '$builder_symbol'.\n");
+                    return;
+                }
+
+                var set_builder = (BuilderFunction) function;
+                set_builder (this.builder);
+            }
+
+            void* function;
+            this.module.symbol (run_symbol, out function);
+            if (function == null) {
+                stderr.printf (@"Module does not contain symbol '$run_symbol'.\n");
+                return;
+            }
+
+            var run = (RunFunction) function;
+            run ();
+        }
+
+        [CCode (has_target=false)]
+        private delegate void RunFunction ();
+
+        [CCode (has_target=false)]
+        private delegate void BuilderFunction (Gtk.Builder builder);
+
+        private Gtk.Window window = new Gtk.Window () {
+            hide_on_close = true
+        };
```

Comment:
Please make the window for the preview default to 800x600 and a AdwWindow

Reply:
That would bean having a AdwHeaderbar in the demo, right?

Reply:
AdwWindow supports window.title

Reply:
But when you cannot specify the root, because it gets added by the previewer?